File: frontend/src/api/teacher.ts
import { apiFetch, apiFetchVoid, teacherHeaders } from "@/api/client";

export interface QueueItem {
  queue_id: number;
  submission_id: number;
  problem_title: string;
  problem_content: string;
  problem_answer: string;
  ocr_raw_text: string | null;
  student_answer: string;
  student_name: string;
  student_id: string | null;
  input_type: string;
  image_path: string | null;
  ai_score: number;
  ai_feedback: string;
  trust_score: number;
  trust_level: "high" | "low";
  queue_type: "score_only" | "full_review";
  sla_deadline: string;
  queued_at: string;
  hallucination_status?: string;
  hallucination_score?: number | null;
  hallucination_issues?: string | null;
  feedback_status?: string;
  solution_status?: string | null;
}

export interface TeacherQueueResponse {
  queue: QueueItem[];
  total: number;
}

export type TeacherAction = "approve" | "modify" | "reject";

export interface TeacherActionRequest {
  action: TeacherAction;
  teacher_score?: number;
  teacher_feedback?: string;
}

export interface TeacherActionResponse {
  queue_id: number;
  action: string;
  reviewed_at: string;
}

export interface FeedbackSummary {
  total_reviewed: number;
  approved: number;
  modified: number;
  rejected: number;
  approval_rate: number;
  avg_score_delta: number;
  low_trust_detection_precision: number;
}

export interface SubmissionOverviewItem {
  submission_id: number;
  problem_id: number;
  problem_title: string;
  student_name: string;
  student_id: string | null;
  input_type: string;
  image_path: string | null;
  student_answer: string | null;
  status: string;
  ai_score: number | null;
  final_score: number | null;
  trust_level: string | null;
  feedback_status: string | null;
  solution_status: string | null;
  submitted_at: string;
  reviewed_at: string | null;
}

export interface SubmissionOverviewResponse {
  submissions: SubmissionOverviewItem[];
  total: number;
  page: number;
  page_size: number;
}

function authHeaders(password: string): HeadersInit {
  return teacherHeaders(true, password);
}

export async function getTeacherQueue(
  password: string,
  trustLevel?: string,
): Promise<TeacherQueueResponse> {
  const params = new URLSearchParams();
  if (trustLevel) params.set("trust_level", trustLevel);
  const url = `/teacher/queue${params.toString() ? `?${params.toString()}` : ""}`;
  return apiFetch(url, {
    headers: authHeaders(password),
  }, "큐 조회 실패");
}

export async function getQueue(trustLevel?: string): Promise<TeacherQueueResponse> {
  const params = new URLSearchParams();
  if (trustLevel) params.set("trust_level", trustLevel);
  const url = `/teacher/queue${params.toString() ? `?${params.toString()}` : ""}`;
  return apiFetch(url, { headers: teacherHeaders() }, "큐 조회 실패");
}

export async function submitTeacherAction(
  queueId: number,
  payload: TeacherActionRequest,
  password: string
): Promise<TeacherActionResponse> {
  return apiFetch(`/teacher/queue/${queueId}/action`, {
    method: "POST",
    headers: authHeaders(password),
    body: JSON.stringify(payload),
  }, "액션 제출 실패");
}

export async function postTeacherAction(
  queueId: number,
  payload: TeacherActionRequest,
): Promise<TeacherActionResponse> {
  const pw = localStorage.getItem("argus_teacher_pw") ?? "";
  return submitTeacherAction(queueId, payload, pw);
}

export async function getFeedbackSummary(
  password: string
): Promise<FeedbackSummary> {
  return apiFetch("/teacher/feedback/summary", {
    headers: authHeaders(password),
  }, "피드백 현황 조회 실패");
}

export async function getSubmissions(params: {
  problem_id?: number;
  status?: string;
  student_name?: string;
  page?: number;
  page_size?: number;
}): Promise<SubmissionOverviewResponse> {
  const query = new URLSearchParams();
  if (params.problem_id) query.set("problem_id", String(params.problem_id));
  if (params.status) query.set("status", params.status);
  if (params.student_name) query.set("student_name", params.student_name);
  if (params.page) query.set("page", String(params.page));
  if (params.page_size) query.set("page_size", String(params.page_size));
  const url = `/teacher/submissions?${query.toString()}`;
  return apiFetch(url, { headers: teacherHeaders() }, "제출 현황 조회 실패");
}

export async function getProblemSubmissions(
  problemId: number,
  params?: { page?: number; page_size?: number },
): Promise<SubmissionOverviewResponse> {
  const query = new URLSearchParams();
  if (params?.page) query.set("page", String(params.page));
  if (params?.page_size) query.set("page_size", String(params.page_size));
  const url = `/teacher/problems/${problemId}/submissions?${query.toString()}`;
  return apiFetch(url, { headers: teacherHeaders() }, "제출 목록 조회 실패");
}

// ===== 그룹 관리 =====

export interface GroupMemberItem { student_id: string; student_name: string }
export interface GroupResponse { id: number; name: string; created_at: string; members: GroupMemberItem[] }
export interface GroupListResponse { groups: GroupResponse[] }

export async function getGroups(): Promise<GroupListResponse> {
  return apiFetch("/teacher/groups", { headers: teacherHeaders(false) }, "그룹 목록 조회 실패");
}

export async function createGroup(name: string): Promise<GroupResponse> {
  return apiFetch("/teacher/groups", {
    method: "POST",
    headers: teacherHeaders(),
    body: JSON.stringify({ name }),
  }, "그룹 생성 실패");
}

export async function deleteGroup(groupId: number): Promise<void> {
  return apiFetchVoid(`/teacher/groups/${groupId}`, {
    method: "DELETE",
    headers: teacherHeaders(false),
  }, "그룹 삭제 실패");
}

export async function addGroupMembers(groupId: number, members: GroupMemberItem[]): Promise<void> {
  return apiFetchVoid(`/teacher/groups/${groupId}/members`, {
    method: "POST",
    headers: teacherHeaders(),
    body: JSON.stringify({ members }),
  }, "멤버 추가 실패");
}

export async function removeGroupMember(groupId: number, studentId: string): Promise<void> {
  return apiFetchVoid(`/teacher/groups/${groupId}/members/${encodeURIComponent(studentId)}`, {
    method: "DELETE",
    headers: teacherHeaders(false),
  }, "멤버 제거 실패");
}

// ===== 숙제 관리 =====

export interface HomeworkProblemItem { problem_id: number; problem_title: string }
export interface HomeworkResponse {
  id: number;
  title: string;
  group_id: number | null;
  group_name: string | null;
  due_date: string | null;
  created_at: string;
  problems: HomeworkProblemItem[];
}
export interface HomeworkListResponse { homeworks: HomeworkResponse[] }

export async function getHomeworks(): Promise<HomeworkListResponse> {
  return apiFetch("/teacher/homeworks", { headers: teacherHeaders(false) }, "숙제 목록 조회 실패");
}

export async function createHomework(payload: {
  title: string;
  group_id?: number | null;
  due_date?: string | null;
  problem_ids: number[];
}): Promise<HomeworkResponse> {
  return apiFetch("/teacher/homeworks", {
    method: "POST",
    headers: teacherHeaders(),
    body: JSON.stringify(payload),
  }, "숙제 생성 실패");
}

export async function deleteHomework(homeworkId: number): Promise<void> {
  return apiFetchVoid(`/teacher/homeworks/${homeworkId}`, {
    method: "DELETE",
    headers: teacherHeaders(false),
  }, "숙제 삭제 실패");
}

File: backend/routers/teacher.py
"""
routers/teacher.py — 교사 검토 대시보드.

GET  /api/v1/teacher/queue              — 미처리 큐 목록
POST /api/v1/teacher/queue/{id}/action  — 승인 / 수정 / 거부
GET  /api/v1/teacher/submissions        — 제출 현황 목록
GET  /api/v1/teacher/problems/{id}/submissions — 문제별 제출 목록
"""

import logging
from datetime import datetime, timezone
from typing import Optional

from fastapi import APIRouter, Depends, Header, HTTPException, Query
from sqlalchemy import select, func
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy.orm import selectinload

from config import settings
from db import get_session
from models import Submission, GradingResult, TeacherQueue, FeedbackLog
from services.job_queue import queue_counts
from schemas.teacher import (
    TeacherQueueResponse,
    TeacherQueueItem,
    TeacherActionRequest,
    TeacherActionResponse,
    SubmissionOverviewItem,
    SubmissionOverviewResponse,
)

logger = logging.getLogger(__name__)
router = APIRouter(prefix="/api/v1/teacher", tags=["teacher"])
AUTO_APPROVE_MARKER = "__AUTO_APPROVED_BY_HALLUCINATION__"


def _verify_teacher(x_teacher_password: str = Header(default="")):
    """단순 비밀번호 인증 (MVP 한정). 헤더 없거나 틀리면 401."""
    if not x_teacher_password or x_teacher_password != settings.teacher_password:
        raise HTTPException(status_code=401, detail="교사 인증 실패")


# ── 검토 큐 목록 ────────────────────────────────────────────────

@router.get("/queue", response_model=TeacherQueueResponse)
async def get_queue(
    trust_level: Optional[str] = None,
    review_status: str = Query(
        "pending",
        description="pending|approved|modify|reject|reviewed|all",
    ),
    sort: str = Query("sla", description="sla|latest"),
    page: int = Query(1, ge=1),
    page_size: int = Query(20, ge=1, le=100),
    db: AsyncSession = Depends(get_session),
    _: None = Depends(_verify_teacher),
):
    """교사 검토 큐 조회. 처리 상태/정렬/신뢰도 필터 지원."""
    query = (
        select(TeacherQueue)
        .join(TeacherQueue.submission)
        .join(Submission.grading_result)
        .options(
            selectinload(TeacherQueue.submission).selectinload(Submission.problem),
            selectinload(TeacherQueue.submission).selectinload(Submission.grading_result),
        )
    )
    normalized_status = (review_status or "pending").lower().strip()
    if normalized_status == "pending":
        query = query.where(TeacherQueue.action.is_(None))
    elif normalized_status == "approved":
        query = query.where(TeacherQueue.action == "approve")
    elif normalized_status == "modify":
        query = query.where(TeacherQueue.action == "modify")
    elif normalized_status == "reject":
        query = query.where(TeacherQueue.action == "reject")
    elif normalized_status == "reviewed":
        query = query.where(TeacherQueue.action.is_not(None))
    elif normalized_status != "all":
        raise HTTPException(status_code=400, detail="review_status 값이 올바르지 않습니다")

    if trust_level:
        query = query.where(func.lower(GradingResult.trust_level) == trust_level.lower())

    total_result = await db.execute(select(func.count()).select_from(query.subquery()))
    total = total_result.scalar() or 0

    if sort == "latest":
        ordered = query.order_by(func.coalesce(TeacherQueue.reviewed_at, TeacherQueue.queued_at).desc())
    elif sort == "sla":
        ordered = query.order_by(TeacherQueue.sla_deadline.asc(), TeacherQueue.queued_at.asc())
    else:
        raise HTTPException(status_code=400, detail="sort 값이 올바르지 않습니다")

    result = await db.execute(ordered.offset((page - 1) * page_size).limit(page_size))
    queue_items = result.scalars().all()

    items = []
    for tq in queue_items:
        sub = tq.submission
        gr = sub.grading_result
        if not gr:
            continue
        auto_approved = tq.action == "approve" and tq.teacher_feedback == AUTO_APPROVE_MARKER
        items.append(
            TeacherQueueItem(
                queue_id=tq.id,
                submission_id=sub.id,
                problem_title=sub.problem.title,
                problem_content=sub.problem.content if sub.problem else "",
                problem_answer=sub.problem.answer if sub.problem else "",
                ocr_raw_text=sub.ocr_raw_text,
                student_answer=sub.student_answer,
                student_name=sub.student_name,
                student_id=sub.student_id,
                input_type=sub.input_type or "text",
                image_path=sub.image_path,
                ai_score=gr.ai_score,
                ai_feedback=gr.ai_feedback,
                trust_score=gr.trust_score,
                trust_level=gr.trust_level,
                queue_type=tq.queue_type,
                sla_deadline=tq.sla_deadline,
                queued_at=tq.queued_at,
                hallucination_status=gr.hallucination_status,
                hallucination_score=gr.hallucination_score,
                hallucination_issues=gr.hallucination_issues,
                feedback_status=gr.feedback_status,
                solution_status=gr.solution_status,
                action=tq.action,
                reviewed_at=tq.reviewed_at,
                auto_approved=auto_approved,
            )
        )

    return TeacherQueueResponse(queue=items, total=total, page=page, page_size=page_size)


@router.get("/queue/health")
async def get_queue_health(
    _: None = Depends(_verify_teacher),
):
    queues = await queue_counts()
    return {"queues": queues}


# ── 교사 액션 ────────────────────────────────────────────────

@router.post("/queue/{queue_id}/action", response_model=TeacherActionResponse)
async def submit_action(
    queue_id: int,
    body: TeacherActionRequest,
    db: AsyncSession = Depends(get_session),
    _: None = Depends(_verify_teacher),
):
    """승인 / 수정 / 거부. 묵시적 자동 승인 없음."""
    result = await db.execute(
        select(TeacherQueue)
        .options(
            selectinload(TeacherQueue.submission).selectinload(Submission.grading_result)
        )
        .where(TeacherQueue.id == queue_id)
    )
    tq = result.scalar_one_or_none()
    if not tq:
        raise HTTPException(status_code=404, detail="큐 항목을 찾을 수 없습니다")
    if tq.action is not None:
        raise HTTPException(status_code=409, detail=f"이미 처리된 항목입니다 (action={tq.action})")

    now = datetime.now(tz=timezone.utc)
    tq.action = body.action
    tq.reviewed_at = now

    if body.action == "modify":
        tq.teacher_score = body.teacher_score
        tq.teacher_feedback = body.teacher_feedback

    # submission 상태 업데이트
    sub = tq.submission
    sub.status = "approved" if body.action in ("approve", "modify") else "rejected"

    # feedback_log 저장
    gr = sub.grading_result
    if gr:
        final_score = (
            body.teacher_score if body.action == "modify" else gr.ai_score
        )
        log = FeedbackLog(
            submission_id=sub.id,
            ai_score=gr.ai_score,
            teacher_score=final_score,
            action=body.action,
            trust_score=gr.trust_score,
            trust_level=gr.trust_level,
        )
        db.add(log)

    await db.commit()

    logger.info(
        f"교사 액션 queue_id={queue_id} action={body.action} "
        f"submission_id={sub.id}"
    )

    return TeacherActionResponse(
        queue_id=queue_id,
        action=body.action,
        reviewed_at=now,
    )


# ── 제출 현황 목록 ────────────────────────────────────────────

@router.get("/submissions", response_model=SubmissionOverviewResponse)
async def get_submissions(
    problem_id: Optional[int] = None,
    status: Optional[str] = None,
    student_name: Optional[str] = None,
    page: int = 1,
    page_size: int = 20,
    db: AsyncSession = Depends(get_session),
    _: None = Depends(_verify_teacher),
):
    query = (
        select(Submission)
        .options(
            selectinload(Submission.problem),
            selectinload(Submission.grading_result),
            selectinload(Submission.teacher_queue),
        )
        .order_by(Submission.submitted_at.desc())
    )
    if problem_id:
        query = query.where(Submission.problem_id == problem_id)
    if status:
        query = query.where(Submission.status == status)
    if student_name:
        query = query.where(Submission.student_name.ilike(f"%{student_name}%"))

    total_result = await db.execute(select(func.count()).select_from(query.subquery()))
    total = total_result.scalar() or 0

    query = query.offset((page - 1) * page_size).limit(page_size)
    result = await db.execute(query)
    submissions = result.scalars().all()

    items = []
    for sub in submissions:
        gr = sub.grading_result
        tq = sub.teacher_queue
        final_score = None
        if tq and tq.action == "modify" and tq.teacher_score is not None:
            final_score = tq.teacher_score
        elif gr:
            final_score = gr.ai_score
        items.append(SubmissionOverviewItem(
            submission_id=sub.id,
            problem_id=sub.problem_id,
            problem_title=sub.problem.title if sub.problem else "",
            student_name=sub.student_name,
            student_id=sub.student_id,
            input_type=sub.input_type,
            image_path=sub.image_path,
            student_answer=sub.student_answer,
            status=sub.status,
            ai_score=gr.ai_score if gr else None,
            final_score=final_score,
            trust_level=gr.trust_level if gr else None,
            feedback_status=gr.feedback_status if gr else None,
            solution_status=gr.solution_status if gr else None,
            submitted_at=sub.submitted_at,
            reviewed_at=tq.reviewed_at if tq else None,
        ))
    return SubmissionOverviewResponse(submissions=items, total=total, page=page, page_size=page_size)


@router.get("/problems/{problem_id}/submissions", response_model=SubmissionOverviewResponse)
async def get_problem_submissions(
    problem_id: int,
    page: int = 1,
    page_size: int = 20,
    db: AsyncSession = Depends(get_session),
    _: None = Depends(_verify_teacher),
):
    query = (
        select(Submission)
        .options(
            selectinload(Submission.problem),
            selectinload(Submission.grading_result),
            selectinload(Submission.teacher_queue),
        )
        .where(Submission.problem_id == problem_id)
        .order_by(Submission.submitted_at.desc())
    )

    total_result = await db.execute(select(func.count()).select_from(query.subquery()))
    total = total_result.scalar() or 0

    query = query.offset((page - 1) * page_size).limit(page_size)
    result = await db.execute(query)
    submissions = result.scalars().all()

    items = []
    for sub in submissions:
        gr = sub.grading_result
        tq = sub.teacher_queue
        final_score = None
        if tq and tq.action == "modify" and tq.teacher_score is not None:
            final_score = tq.teacher_score
        elif gr:
            final_score = gr.ai_score
        items.append(SubmissionOverviewItem(
            submission_id=sub.id,
            problem_id=sub.problem_id,
            problem_title=sub.problem.title if sub.problem else "",
            student_name=sub.student_name,
            student_id=sub.student_id,
            input_type=sub.input_type,
            image_path=sub.image_path,
            student_answer=sub.student_answer,
            status=sub.status,
            ai_score=gr.ai_score if gr else None,
            final_score=final_score,
            trust_level=gr.trust_level if gr else None,
            feedback_status=gr.feedback_status if gr else None,
            solution_status=gr.solution_status if gr else None,
            submitted_at=sub.submitted_at,
            reviewed_at=tq.reviewed_at if tq else None,
        ))
    return SubmissionOverviewResponse(submissions=items, total=total, page=page, page_size=page_size)
